mgmt: mcumgr: replace Tinycrypt by PSA

tomi-font opened the pull request:
As part of ongoing work to move away from TinyCrypt and towards PSA (#43712), make `fs_mgmt` use either PSA (when available) or MbedTLS (as a fallback) for SHA-256.
    
The use of PSA is guarded by CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT which requires a locally-built PSA core for devices without TF-M.

tomi-font committed:
mgmt: mcumgr: replace Tinycrypt by PSA

As part of ongoing work to move away from TinyCrypt and towards PSA
(#43712), make fs_mgmt use either PSA (when available) or MbedTLS
(as a fallback) for SHA-256.

The use of PSA is guarded by CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT
which requires a locally-built PSA core for devices without TF-M.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

nordicjm reviewed (changes requested):
I cannot accept this change. I understand that having multiple security backends is not great, but when I enable the existing tinycrypt mode for this feature, I get:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      208648 B       472 KB     43.17%
             RAM:       61280 B       256 KB     23.38%
        IDT_LIST:          0 GB        32 KB      0.00%
```

When I switch over to mbedtls I get:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      219436 B       472 KB     45.40%
             RAM:       61280 B       256 KB     23.38%
        IDT_LIST:          0 GB        32 KB      0.00%
```

This is an increase of 10.6KiB, for a simple sha256 hash. I don't know what the overhead of this is in tfm but the overhead without is not acceptable

carlescufi commented:
Would it be an option to compile this conditionally? i.e. by default mcumgr would use PSA APIs, since that is the direction the project has chosen, but with Kconfig one could still enable Tinycrypt.

tomi-font commented:
In my opinion it could definitely be. @frkv?

nandojve commented:
> Would it be an option to compile this conditionally? i.e. by default mcumgr would use PSA APIs, since that is the direction the project has chosen, but with Kconfig one could still enable Tinycrypt.

+1

ceolin commented:
> 
> This is an increase of 10.6KiB, for a simple sha256 hash. I don't know what the overhead of this is in tfm but the overhead without is not acceptable

@nordicjm what is the target / configuration you tested it ? 10k is really too much, I have seen this building

`west build -p -b  nrf52840dk/nrf52840 tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported/ -- -DEXTRA_CONF_FILE=configuration/sha256.conf`

But lets not jump into early conclusions, I have just disabled some additional features that were enabled by default when building with mbedTLS and the difference was way smaller.

Before any change (building with mbedTLS):
```
-- Zephyr version: 3.6.99 (/home/faceolin/p/zephyrproject/zephyr), build: v3.6.0-2908-gf366bd3fb3f0
[271/271] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       75496 B         1 MB      7.20%
             RAM:       14392 B       256 KB      5.49%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/faceolin/p/zephyrproject/zephyr/build/zephyr/zephyr.elf for board: nrf52840dk
```

After disable unnecessary options for this test:
```
 [265/265] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       65192 B         1 MB      6.22%
             RAM:       14392 B       256 KB      5.49%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/faceolin/p/zephyrproject/zephyr/build/zephyr/zephyr.elf for board: nrf52840dk
```

Without these changes (current upstream)
```
-- Zephyr version: 3.6.99 (/home/faceolin/p/zephyrproject/upstream-main), build: v3.6.0-3023-g7084662cc80b
[181/181] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       64772 B         1 MB      6.18%
             RAM:       14520 B       256 KB      5.54%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/faceolin/p/zephyrproject/upstream-main/build/zephyr/zephyr.elf for board: nrf52840dk
```

I don't think supporting TinyCrypt indefinitely is a proper solution (obviously, we don't want to be disruptive in a LTS release). We need to have a proper solution. 

I understand the concern with additional resources needed, but there are also multiple benefits adopting it.

nordicjm commented:
Tinycrypt itself I'm not so bothered about, the 10KiB jump is what I'm bothered about. In tinycrypt you enable it then have to enable what features you want, seemingly in mbedtls you enable it then have to go and disable features you don't want, that's not really useful and certainly not what other Kconfigs in zephyr subsystems do, you enable a feature with a set of minimal defaults then enable the additional ones you want, so if the size difference from that can be vastly reduced by having minimal defaults and a way for features to select what they need e.g. SHA256 which enables them then that's a path on the right track

nordicjm commented:
Build string:
```
cmake -GNinja -DBOARD=nrf52840dk/nrf52840 -DOVERLAY_CONFIG="overlay-bt.conf" -DCONFIG_TINYCRYPT=y -DCONFIG_TINYCRYPT_SHA256=y -DCONFIG_MCUMGR_GRP_FS_CHECKSUM_HASH=y -DCONFIG_MCUMGR_GRP_FS_HASH_SHA256=y ..
```

ceolin commented:
> Tinycrypt itself I'm not so bothered about, the 10KiB jump is what I'm bothered about. In tinycrypt you enable it then have to enable what features you want, seemingly in mbedtls you enable it then have to go and disable features you don't want, that's not really useful and certainly not what other Kconfigs in zephyr subsystems do, you enable a feature with a set of minimal defaults then enable the additional ones you want, so if the size difference from that can be vastly reduced by having minimal defaults and a way for features to select what they need e.g. SHA256 which enables them then that's a path on the right track

I agree with you, it is non-sense pulling a bunch of functionalities that were not requested. My comment was to address the size increase issue. There will *much likely* be an increase with PSA but that must not be in this magnitude.
Regarding TinyCrypt, everyone should care about it since that is big security problem for the project.

tomi-font commented:
With the changes from https://github.com/zephyrproject-rtos/zephyr/pull/72078 that removes the default enabling of hash algorithms for MbedTLS, we get to the following differences when building `smp_svr` with `overlay-bt.conf` for `nrf52840dk/nrf52840`:

TinyCrypt (reference):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      208428 B       472 KB     43.12%
             RAM:       61248 B       256 KB     23.36%
        IDT_LIST:          0 GB        32 KB      0.00%
```

MbedTLS (+480 bytes):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      208908 B       472 KB     43.22%
             RAM:       61248 B       256 KB     23.36%
        IDT_LIST:          0 GB        32 KB      0.00%
```

And when using directly `mbedtls_sha256` instead of `mbedtls_md` (which is now the case with my latest changes):

MbedTLS (+68 bytes):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      208496 B       472 KB     43.14%
             RAM:       61248 B       256 KB     23.36%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Additionally, if compiling with `overlay-fs.conf` instead of `overlay-bt.conf` the difference shrinks a bit further. This is because currently the bluetooth subsystem pulls in TinyCrypt unconditionally.

Now, regarding the PSA case. As of now the PSA API will only be used when building with TF-M, in which case the crypto operations happen in TF-M. Because TF-M already has the code, the net difference in code size when going from TinyCrypt to PSA is negative.

Building with `overlay-fs.conf` for `nrf5340dk/nrf5340/cpuapp/ns` (to have TF-M):

TinyCrypt (reference):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       79540 B       192 KB     40.46%
             RAM:       16944 B       192 KB      8.62%
        IDT_LIST:          0 GB        32 KB      0.00%
```

PSA (-768 bytes):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       78772 B       192 KB     40.07%
             RAM:       16944 B       192 KB      8.62%
        IDT_LIST:          0 GB        32 KB      0.00%
```

tomi-font committed:
mbedtls: introduce PSA-specific Kconfig file

It is meant specifically for configuration of the PSA crypto library.

The underlying PSA configuration items are guarded by the condition
that a PSA crypto provider must be present, which is the case when
either TF-M is in use or MbedTLS's PSA core is built as part of
the application image.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
tests: mgmt: mcumgr: fs_mgmt_hash_supported: make CI green

Exclude some more platforms from the tests.
They provoke devicetree-related build errors which weren't introduced
by the changes in this PR (#71947).

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
This is ready for review. (And depends on https://github.com/zephyrproject-rtos/zephyr/pull/72078, so don't mind the 3 first commits, or react on that PR.)

tomi-font commented:
With the latest changes I made the PSA implementation is also used in the case where `CONFIG_MBEDTLS_PSA_CRYPTO_C` is enabled.
This is (typically) not the case by default and also requires `CONFIG_MBEDTLS_ENTROPY_ENABLED` to be enabled.

When building (`smp_svr` with `overlay-fs.conf` for `nrf52840dk/nrf52840`) with those Kconfig options enabled, the net difference is +1368 bytes when using the PSA implementation (compared to the MbedTLS one).
This is caused by what gets pulled in when enabling `CONFIG_MBEDTLS_PSA_CRYPTO_C`; for instance (and most likely the biggest culprit here) the DRBG defaults to the AES-256-based one, so AES-256 gets automatically pulled in.

This will be improved (e.g. by defaulting to using the HMAC-based DRBG and/or reducing the number of Kconfig options enabled by default as was done with the hash algorithms for this PR) as part of upcoming work.

tomi-font commented:
ping @ceolin @de-nordic @nordicjm @SebastianBoe

ceolin commented on subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt_hash_checksum_sha256.c:
```diff
@@ -13,90 +13,46 @@
-#include <tinycrypt/sha256.h>
+#define PSA_IMPLEMENTATION IS_ENABLED(CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT)
+
+#if PSA_IMPLEMENTATION
```
Why don't use `#ifdef CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT`  directly ? It is easier to see what is the build option

tomi-font replied:
I thought that `PSA_IMPLEMENTATION` would be more self-explanatory, also because it is repeated in this source file. But I can change it to only using `CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT`.

nordicjm replied:
the ifdef looks cleaner

ceolin commented on subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt_hash_checksum_sha256.c:
```diff
@@ -108,30 +64,41 @@ static int fs_mgmt_hash_checksum_sha256(struct fs_file_t *file, uint8_t *output,
+	psa_hash_abort(&psa_hash_ctx);
+#else
+	mbedtls_sha256_free(&mbed_hash_ctx);
+#endif
```
looking all these ifdefs, wouldn't be cleaner having two distinct implementations ? like currently exists for TinyCrypt and mbedTLS ?

tomi-font replied:
The implementations have so much in common that I thought it would be a bit of a pity to duplicate the code. (And the `#ifdef`s are partly the reason why I introduced the `PSA_IMPLEMENTATION` macro, to make them more readable.)

What do the others think?

de-nordic replied:
I think that, by the signatures of the functions, we could just define "API" here in form:
```
typedef  ...  hash_ctx // this would define psa_hash_operation_t or mbedtls_sha256_context type object
int hash_setup(hash_ctx *ctx) // optionally taking hash type, but I would not do that, this is local thing.
int hash_update(hash_ctx *ctx, uint8_t *buff, size_t buf_size)
int hash_finish(hash_ctx *ctx, uint8_t *buff)
int hash_cleanup(hash_ctx *ctx) // serves as abort/free
```

Group above in `#ifdef CONFIG_MBEDTLS_PSA_CRYPTO_CLIEN, #else` block above or below `fs_mgmt_hash_checksum_sha256` and cut-down `#ifdefs` in that function. 

IMHO.

ceolin reviewed (commented):
The changes look correct to me, just some comments about aesthetics  of the changes.
Other than that is +1 for the mbedTLS/PSA changes.

nordicjm commented on tests/subsys/mgmt/mcumgr/all_options/prj.conf:
```diff
@@ -4,8 +4,6 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 CONFIG_ZTEST=y
-CONFIG_TINYCRYPT=y
-CONFIG_TINYCRYPT_SHA256=y
```
this needs to have mbedtls enabled, as the name implies this test ensures all options can be enabled

nordicjm commented on tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported/configuration/crc32.conf:
```diff
@@ -4,3 +4,4 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 CONFIG_MCUMGR_GRP_FS_CHECKSUM_IEEE_CRC32=y
+CONFIG_MCUMGR_GRP_FS_HASH_SHA256=n
```
this is not needed, not sure why it's here?

tomi-font replied:
This is out of consistency with `sha256.conf`.
If CRC32 is enabled (which it is by default) it will be set as the default (`MCUMGR_GRP_FS_CHECKSUM_HASH_DEFAULT` in `fs_mgmt.c`) so it felt like explicitly disabling it in the case of SHA-256 was a good idea.
Just doing the same thing the other way around here even though as of now it's not necessary.

nordicjm commented on tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported/configuration/all.conf:
```diff
@@ -3,7 +3,8 @@
-CONFIG_TINYCRYPT_SHA256=y
 CONFIG_MCUMGR_GRP_FS_CHECKSUM_IEEE_CRC32=y
 CONFIG_MCUMGR_GRP_FS_HASH_SHA256=y
+
```
```suggestion
```

nordicjm commented on tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported/configuration/sha256.conf:
```diff
@@ -3,6 +3,8 @@
-CONFIG_TINYCRYPT_SHA256=y
+CONFIG_MCUMGR_GRP_FS_CHECKSUM_IEEE_CRC32=n
 CONFIG_MCUMGR_GRP_FS_HASH_SHA256=y
+
```
```suggestion
```

nordicjm reviewed (changes requested):
Changes look good, thanks for improving and getting the size down. Some minor nits.

de-nordic reviewed (dismissed):
Looks good, thanks!

tomi-font commented:
Rebased and resolved conflicts.

tomi-font commented:
All CI failures should be addressed now.
